3534 Password Toggle Text Variation

bgoewert opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

https://core.trac.wordpress.org/ticket/3534

This a text variation of #3725 to address accessibility concerns and add support for various password manager implementations. The button is separated from the input while still keeping it inline — allowing for password managers which add icons directly to the input, in contrast to those that insert another element to add their respective icon. I did not include the icon, which is inconsistent with the other toggle buttons, so as to provide maximum input space on the setup-config.php form. But the icon can easily be added back to remain consistent.

![separated_pwd-toggle_inactive](https://i.imgur.com/IM3Alya.png)
_Figure 1: Separated text password toggle w/ LastPass extension - inactive_

![separated_pwd-toggle_inactive](https://i.imgur.com/MoDH3EV.png)
_Figure 2: Separated text password toggle w/ LastPass extension  - active_

As a consequence of adding `display: flex` to `.wp-pwd`, the layout for install.php and user-new.php was also affected. I've also added changes to address this. For user-edit.php, [user-profile.js#L232](https://core.trac.wordpress.org/browser/trunk/src/js/_enqueues/admin/user-profile.js#L232) applies `display: block` directly to `.wp-pwd` and was not affected.

![install_before_change](https://i.imgur.com/Ub3imZV.png)
_Figure 3: install.php before._

![install_before_fix](https://i.imgur.com/3h0p8ct.png)
_Figure 4: install.php after change._

![install_after_fix](https://i.imgur.com/KQRCHTp.png)
_Figure 5: install.php after fix._

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Commit:
Add password visibility switch to setup-config

Commit:
Fix position differences between browsers

Commit:
Remove aria show/hide label

Commit:
Add padding so password visibility button does not overlap for rtl/ltr text

Commit:
Change the button width back to 40px to be consistent

Commit:
Change .wp-pwd back to inline-block to stay consistent

Commit:
Remove leftover `display: initial` from forms.css

Commit:
Rename setup-config.js to password-toggle.js

bgoewert committed:
Update password-toggle.js @output to new file name

Need to use global search next time when renaming things.

Commit:
Update #pwd-toggle to .pwd-toggle

Commit:
Update `password-toggle.js` to toggle any and all `.pwd-toggle` fields

Commit:
Change default aria-label from "Hide" to "Show"

Commit:
Reword file description for password-toggle.js

Commit:
Fix spacing inside parenthesis so it's consistent

Commit:
Change `toggle` to `toggleElements` to imply potentially several `.pwd-toggle` elements

Commit:
Change `.wp-pwd` from inline to block

Commit:
Fix password-toggle.js so it correctly toggles the input that the button applies to if there are multiple fields

Commit:
Add `hide-if-no-js` to `setup-config.php` toggle as it is unusable w/o JS

Commit:
Update forms.css and install.css for appropriate input padding

Commit:
Remove `data-pw` from setup-config.php

Commit:
Change `aria-label` on toggle including translation support

Commit:
Change install.css selector for `padding-right` so as to not affect installation page

Commit:
Change `margin-top` to not affect top margin on profile page

Commit:
Change icon to text and fix pw mgr icon support

Commit:
Merge branch 'trunk'

Commit:
Remove unneeded `margin: 0` from `.pwd-toggle`. Margin is already set to 0 in `.wp-core-ui .button`.

Commit:
Fix password strength result and button height

Commit:
Remove unnecessary `.pwd-toggle` styles from forms.css

Commit:
Re-add icon to button for consistency

Commit:
Increase input size slightly to allow for pwd-toggle. Also fixes orphan words in the input descriptions.

Commit:
Merge branch 'trunk'

Conflicts:
	src/wp-admin/install.php
	src/wp-admin/user-new.php

Commit:
Revert `.pwd-toggle` to default button width/padding

Commit:
Fix input width on narrow screens and update `calc()` based on reversion of width/padding

audrasjb commented:
Hello, there is a small conflict in steup-config.php since I committed [[55110]](https://core.trac.wordpress.org/changeset/55110)

Commit:
Merge branch 'trunk'

Conflicts:
	src/wp-admin/setup-config.php

bgoewert commented:
Thanks @audrasjb. I just merged trunk to address the conflict.

SergeyBiryukov commented on Gruntfile.js:
```diff
@@ -275,6 +275,7 @@ module.exports = function(grunt) {
 					[ WORKING_DIR + 'wp-admin/js/plugin-install.js' ]: [ './src/js/_enqueues/admin/plugin-install.js' ],
 					[ WORKING_DIR + 'wp-admin/js/post.js' ]: [ './src/js/_enqueues/admin/post.js' ],
 					[ WORKING_DIR + 'wp-admin/js/postbox.js' ]: [ './src/js/_enqueues/admin/postbox.js' ],
+					[ WORKING_DIR + 'wp-admin/js/password-toggle.js' ]: [ './src/js/_enqueues/admin/password-toggle.js' ],
```
Minor nitpick, could this be moved directly after `wp-admin/js/password-strength-meter.js`, to line 275, for alphabetical order?

bgoewert replied:
Should have realized these were alphabetical. I'll reorder.

SergeyBiryukov commented on src/wp-includes/script-loader.php:
```diff
@@ -1180,6 +1180,9 @@ function wp_default_scripts( $scripts ) {
 		)
 	);
 
+	$scripts->add( 'password-toggle', "/wp-admin/js/password-toggle$suffix.js", array(), false, 1 );
+	$scripts->set_translations( 'password-toggle' );
+
```
Minor nitpick, could this block be moved after `$scripts->set_translations( 'password-strength-meter' )`, to line 1165?

bgoewert replied:
Will do. Also noticed that `wp-i18n` is not listed as a dependency, even though I referenced it in `password-toggle.js`. Am I correct to think that this needs to be added as a dep?

peterwilsoncc replied:
> Also noticed that `wp-i18n` is not listed as a dependency, even though I referenced it in `password-toggle.js`.

The `set_translations()` method ensures it's included in the dependencies so you don't need to include it here.

https://github.com/WordPress/wordpress-develop/blob/6036ee03af6aba166ab86b0dd9357dff2e2128da/src/wp-includes/class-wp-scripts.php#L576-L589

Commit:
Reorder `password-toggle.js` in gruntfile so it is listed alphabetically.

Commit:
Update `script-loader.php` to group `password-toggle` near `password-strength-meter`

Commit:
Remove `display: flex` from forms.css that was causing the `wp-login.php` toggle from aligning properly.

Commit:
Add fixed width to toggle button

Commit:
Add `.password-input-wrapper` with `display: inline-block` to forms.css.
This keep toggle buttons in-line, which was removed with the removal of `display: flex`.

Commit:
Merge branch 'trunk'

Conflicts:
	src/wp-admin/install.php

sabernhardt commented on src/wp-admin/css/forms.css:
```diff
@@ -551,6 +551,17 @@ fieldset label,
+
+.wp-pwd button {
+	height: min-content;
+	width: 4.875rem;
```
Fixed width buttons can be trouble in other languages.

Spanish overflows a little:
![setup-config in Spanish](https://user-images.githubusercontent.com/17100257/216280497-7ef9fd18-c29c-478a-aaf9-29db0b1c0153.png)

Malayalam overflows a lot:
![setup-config in Malayalam](https://user-images.githubusercontent.com/17100257/216280495-725e144c-a485-4c11-b310-2aa8dc8835ad.png)

sabernhardt commented on src/wp-admin/css/forms.css:
```diff
@@ -551,6 +551,17 @@ fieldset label,
+
+.wp-pwd button.pwd-toggle .dashicons {
+	position: relative;
+	top: 0.25rem
```
Missing semicolon.

sabernhardt commented on src/wp-admin/css/forms.css:
```diff
@@ -593,6 +604,10 @@ fieldset label,
 }
 
+.password-input-wrapper {
+	display: inline-block;
```
This should not be `inline-block` at smaller sizes. It seems fine on `setup-config.php`, but it changes both the Installation and Profile screens.

Installation:
![installation screen at 500 pixels wide](https://user-images.githubusercontent.com/17100257/216281699-7c5edb75-908e-4827-8523-c6b702cec23b.png)

Profile:
![profile screen at 500 pixels wide](https://user-images.githubusercontent.com/17100257/216281702-b2a39fec-cd3b-415f-8fd7-d93b6471e7fa.png)

sabernhardt commented on src/wp-admin/install.php:
```diff
@@ -140,12 +140,14 @@ function display_setup_form( $error = null ) {
 				<div class="wp-pwd">
 					<?php $initial_password = isset( $_POST['admin_password'] ) ? stripslashes( $_POST['admin_password'] ) : wp_generate_password( 18 ); ?>
-					<input type="password" name="admin_password" id="pass1" class="regular-text" autocomplete="new-password" spellcheck="false" data-reveal="1" data-pw="<?php echo esc_attr( $initial_password ); ?>" aria-describedby="pass-strength-result" />
+					<span class="password-input-wrapper">
```
A `div` could be set to `inline-block`, if appropriate, to avoid putting the password-strength `div` inside a `span`.

sabernhardt reviewed (changes requested):
Thanks for adding text to the button. I marked some specific lines in the files.

This PR still adds an icon-only button to the `setup-config` page on smaller screens, which would need the `2.5rem` side padding behind the button (both on the right, even in RTL).

English:
![Password field in English at 500 pixels wide](https://user-images.githubusercontent.com/17100257/216288885-a25550f9-c038-4415-ac04-c3974195d630.png)

Arabic:
![Password field in Arabic at 500 pixels wide](https://user-images.githubusercontent.com/17100257/216288892-fe0df0e9-581e-4167-bd39-e0b65c6f5264.png)

bgoewert commented:
Closed in [r56008](https://core.trac.wordpress.org/changeset/56008)
